[Fix #13395] False positive for `Lint/UselessAssignment` when assigning in branch and block

pCosta99 opened the pull request:
As demonstrated in the issue #13395, there is a false positive as such:
```ruby
changed = false
^^^^^^^ Useless assignment to variable - `changed`.

if Random.rand > 1
  changed = true
end

[].each do
  changed = true
end

puts changed
```

In the presence of a conditional, the cop `Lint/UselessAssignment` can't be certain if the variable will be reassigned or not, so, very much like the case for blocks, the best approach is simply to ignore it.

There appears to be no benefit in actually looking into all branches and checking if all branches reassign since even then the outer assignment defines the variable.

This also applies to case statements.

-----------------

Before submitting the PR make sure the following are checked:

* [X] The PR relates to *only* one subject with a clear title and description in grammatically correct, complete sentences.
* [X] Wrote [good commit messages][1].
* [X] Commit message starts with `[Fix #issue-number]` (if the related issue exists).
* [X] Feature branch is up-to-date with `master` (if not - rebase it).
* [X] Squashed related commits together.
* [X] Added tests.
* [X] Ran `bundle exec rake default`. It executes all tests and runs RuboCop on its own code.
* [X] Added an entry (file) to the [changelog folder](https://github.com/rubocop/rubocop/blob/master/changelog/) named `{change_type}_{change_description}.md` if the new code introduces user-observable changes. See [changelog entry format](https://github.com/rubocop/rubocop/blob/master/CONTRIBUTING.md#changelog-entry-format) for details.

[1]: https://chris.beams.io/posts/git-commit/

pCosta99 commented on lib/rubocop/cop/variable_force/variable.rb:
```diff
@@ -31,11 +33,21 @@ def initialize(name, declaration_node, scope)
 
+        def mark_last_as_reassigned!(assignment)
+          return if captured_by_block?
+          return if candidate_condition?(assignment.node.parent)
```
@koic Can I make `parent` an attribute in `RuboCop::Cop::VariableForce::Assignment`? 
Seems like a worthy refactor, makes sense as an attribute and simplifies a bunch of accesses throughout the class

bbatsov commented:
> There appears to be no benefit in actually looking into all branches and checking if all branches reassign since even then the outer assignment defines the variable.

I don't quite get this part, as by looking at them you can actually determine whether some assignment is void or not.

pCosta99 commented:
> > There appears to be no benefit in actually looking into all branches and checking if all branches reassign since even then the outer assignment defines the variable.
> 
> I don't quite get this part, as by looking at them you can actually determine whether some assignment is void or not.

Can you create an example? What I was trying to highlight is situations like this:

```ruby
changed = false

if Random.rand > 1
  changed = true
else 
  changed = false
end

[].each do
  changed = true
end

puts changed
```

Even considering that any branch will assign to `changed` the assignment above is not useless since it creates the variable. 
Out of the top of my head no examples come to mind that it matters to check all branches.

bbatsov commented:
> Even considering that any branch will assign to changed the assignment above is not useless since it creates the variable.

So, your point applies only if there's a usage outside the branches, right? (in your case the `puts`)

pCosta99 commented:
> > Even considering that any branch will assign to changed the assignment above is not useless since it creates the variable.
> 
> So, your point applies only if there's a usage outside the branches, right? (in your case the `puts`)

I'd say so. And testing with, for instance:
```ruby
def some_method
  changed = false
  ^^^^^^^ Useless assignment to variable - `changed`.

  if Random.rand > 1
    changed = true
    ^^^^^^^ Useless assignment to variable - `changed`.
  else
    changed = false
    ^^^^^^^ Useless assignment to variable - `changed`.
  end
end
```

works as expected. However, one important detail, in the following scenario:

```ruby
def some_method
  changed = false

  if Random.rand > 1
    changed = true
  end

  [].each do
    changed = true
  end
end
```

the variable is not unused since it is captured by a block. This is respectful of the comments under `RuboCop::Cop::VariableForce::Variable::used?` that quote the following:

>         # This is a convenient way to check whether the variable is used
>        # in its entire variable lifetime.
>        # For more precise usage check, refer Assignment#used?.
>        #
>        # Once the variable is captured by a block, we have no idea
>        # when, where, and how many times the block would be invoked.
>        # This means we cannot track the usage of the variable.
>       # So we consider it's used to suppress false positive offenses.

@bbatsov

pCosta99 commented:
It's been a long time, I've kinda forgotten about this, sorry, where do we stand on this @bbatsov ?

navels commented:
This is still needed . . . thanks

pCosta99 commented:
@navels not sure how to proceed here to be honest

koic commented:
Let’s merge this proposal for now and proceed with resolving the issue. The different tasks can be addressed separately.

navels commented:
Thanks!
